File: Tween/Setting/SettingBase.vb
' Tween - Client of Twitter
' Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri_feather@gmail.com>
'           (c) 2008-2011 Moz (@syo68k)
'           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
'           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
'           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
' All rights reserved.
' 
' This file is part of Tween.
' 
' This program is free software; you can redistribute it and/or modify it
' under the terms of the GNU General Public License as published by the Free
' Software Foundation; either version 3 of the License, or (at your option)
' any later version.
' 
' This program is distributed in the hope that it will be useful, but
' WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
' or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
' for more details. 
' 
' You should have received a copy of the GNU General Public License along
' with this program. If not, see <http://www.gnu.org/licenses/>, or write to
' the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
' Boston, MA 02110-1301, USA.

Public MustInherit Class SettingBase(Of T As {Class, New})

    Private Shared lockObj As New Object

    Protected Shared Function LoadSettings(ByVal FileId As String) As T
        Try
            SyncLock lockObj
                Using fs As New IO.FileStream(GetSettingFilePath(FileId), IO.FileMode.Open)
                    fs.Position = 0
                    Dim xs As New Xml.Serialization.XmlSerializer(GetType(T))
                    Dim instance As T = DirectCast(xs.Deserialize(fs), T)
                    fs.Close()
                    Return instance
                End Using
            End SyncLock
        Catch ex As System.IO.FileNotFoundException
            Return New T()
        Catch ex As Exception
            Dim backupFile As String = IO.Path.Combine( _
                    IO.Path.Combine( _
                        My.Application.Info.DirectoryPath, _
                        "TweenBackup1st"), _
                    GetType(T).Name + FileId + ".xml")
            If IO.File.Exists(backupFile) Then
                Try
                    SyncLock lockObj
                        Using fs As New IO.FileStream(backupFile, IO.FileMode.Open)
                            fs.Position = 0
                            Dim xs As New Xml.Serialization.XmlSerializer(GetType(T))
                            Dim instance As T = DirectCast(xs.Deserialize(fs), T)
                            fs.Close()
                            MessageBox.Show("File: " + GetSettingFilePath(FileId) + Environment.NewLine + "Use old setting file, because application can't read this setting file.")
                            Return instance
                        End Using
                    End SyncLock
                Catch ex2 As Exception
                End Try
            End If
            MessageBox.Show("File: " + GetSettingFilePath(FileId) + Environment.NewLine + "Use default setting, because application can't read this setting file.")
            Return New T()
            'ex.Data.Add("FilePath", GetSettingFilePath(FileId))
            'Dim fi As New IO.FileInfo(GetSettingFilePath(FileId))
            'ex.Data.Add("FileSize", fi.Length.ToString())
            'ex.Data.Add("FileData", IO.File.ReadAllText(GetSettingFilePath(FileId)))
            'Throw
        End Try
    End Function

    Protected Shared Function LoadSettings() As T
        Return LoadSettings("")
    End Function

    Protected Shared Sub SaveSettings(ByVal Instance As T, ByVal FileId As String)
        Dim cnt As Integer = 0
        Dim err As Boolean = False
        Dim fileName As String = GetSettingFilePath(FileId)
        If Instance Is Nothing Then Exit Sub
        Do
            err = False
            cnt += 1
            Try
                SyncLock lockObj
                    Using fs As New IO.FileStream(fileName, IO.FileMode.Create)
                        fs.Position = 0
                        Dim xs As New Xml.Serialization.XmlSerializer(GetType(T))
                        xs.Serialize(fs, Instance)
                        fs.Flush()
                        fs.Close()
                    End Using
                    Dim fi As New IO.FileInfo(fileName)
                    If fi.Length = 0 Then
                        If cnt > 3 Then
                            Throw New Exception
                            Exit Sub
                        End If
                        Threading.Thread.Sleep(1000)
                        err = True
                    End If
                End SyncLock
            Catch ex As Exception
                '検証エラー or 書き込みエラー
                If cnt > 3 Then
                    'リトライオーバー
                    MessageBox.Show("Can't write setting XML.(" + fileName + ")", "Save Settings", MessageBoxButtons.OK)
                    'Throw New System.InvalidOperationException("Can't write setting XML.(" + fileName + ")")
                    Exit Sub
                End If
                'リトライ
                Threading.Thread.Sleep(1000)
                err = True
            End Try
        Loop While err
    End Sub

    Protected Shared Sub SaveSettings(ByVal Instance As T)
        SaveSettings(Instance, "")
    End Sub

    Public Shared Function GetSettingFilePath(ByVal FileId As String) As String
        Return IO.Path.Combine(My.Application.Info.DirectoryPath, GetType(T).Name + FileId + ".xml")
    End Function
End Class

File: OpenTween/Setting/SettingBase.cs
// OpenTween - Client of Twitter
// Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri.feather@gmail.com>
//           (c) 2008-2011 Moz (@syo68k)
//           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
//           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
//           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
//           (c) 2011      kim_upsilon (@kim_upsilon) <https://upsilo.net/~upsilon/>
// All rights reserved.
//
// This file is part of OpenTween.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>, or write to
// the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
// Boston, MA 02110-1301, USA.

#nullable enable

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Windows.Forms;
using System.Xml;
using System.Xml.Serialization;

namespace OpenTween
{
    public abstract class SettingBase<T>
        where T : class, new()
    {
        /// <summary>
        /// XML に含まれる追加データを保持するフィールド
        /// </summary>
        /// <remarks>
        /// 他バージョンから設定をコピーした場合など、クラスに存在しないフィールドが
        /// XML に含まれていた場合に破棄せず保持するため必要となる。
        /// </remarks>
        [XmlAnyElement]
        public XmlElement[] ExtraElements = Array.Empty<XmlElement>();

        private static readonly object LockObj = new();

        protected static T LoadSettings(string settingsPath, string fileId)
        {
            try
            {
                var settingFilePath = GetSettingFilePath(settingsPath, fileId);
                if (!File.Exists(settingFilePath))
                {
                    return new T();
                }

                lock (LockObj)
                {
                    using var fs = new FileStream(settingFilePath, FileMode.Open, FileAccess.Read);
                    fs.Position = 0;
                    var xs = new XmlSerializer(typeof(T));
                    var instance = (T)xs.Deserialize(fs);
                    return instance;
                }
            }
            catch (FileNotFoundException)
            {
                return new T();
            }
            catch (Exception)
            {
                var backupFile = Path.Combine(
                        Path.Combine(
                            Application.StartupPath,
                            ApplicationSettings.AssemblyName + "Backup1st"),
                        typeof(T).Name + fileId + ".xml");
                if (File.Exists(backupFile))
                {
                    try
                    {
                        lock (LockObj)
                        {
                            using var fs = new FileStream(backupFile, FileMode.Open, FileAccess.Read);
                            fs.Position = 0;
                            var xs = new XmlSerializer(typeof(T));
                            var instance = (T)xs.Deserialize(fs);
                            MessageBox.Show("File: " + GetSettingFilePath(settingsPath, fileId) + Environment.NewLine + "Use old setting file, because application can't read this setting file.");
                            return instance;
                        }
                    }
                    catch (Exception)
                    {
                    }
                }
                MessageBox.Show("File: " + GetSettingFilePath(settingsPath, fileId) + Environment.NewLine + "Use default setting, because application can't read this setting file.");
                return new T();
            }
        }

        protected static T LoadSettings(string settingsPath)
            => LoadSettings(settingsPath, "");

        protected static void SaveSettings(T instance, string settingsPath, string fileId)
        {
            const int SaveRetryMax = 3;

            if (instance == null)
                return;

            var retryCount = 0;
            Exception? lastException = null;

            var filePath = GetSettingFilePath(settingsPath, fileId);
            do
            {
                var tmpfilePath = GetSettingFilePath(settingsPath, "_" + Path.GetRandomFileName());
                try
                {
                    lock (LockObj)
                    {
                        using (var stream = new FileStream(tmpfilePath, FileMode.Create, FileAccess.Write))
                        {
                            var serializer = new XmlSerializer(typeof(T));
                            serializer.Serialize(stream, instance);
                            stream.Flush();
                        }

                        var fileInfo = new FileInfo(tmpfilePath);
                        if (fileInfo.Length != 0)
                        {
                            // 成功
                            File.Copy(tmpfilePath, filePath, true);
                            return;
                        }
                    }
                }
                catch (Exception e)
                {
                    lastException = e;
                }
                finally
                {
                    try
                    {
                        if (File.Exists(tmpfilePath))
                            File.Delete(tmpfilePath);
                    }
                    catch (Exception)
                    {
                    }
                }

                // リトライ
                retryCount++;
                Thread.Sleep(1000);
            }
            while (retryCount <= SaveRetryMax);

            // リトライオーバー
            if (lastException != null)
                MyCommon.ExceptionOut(lastException);

            MessageBox.Show("Can't write setting XML.(" + filePath + ")", "Save Settings", MessageBoxButtons.OK);
        }

        protected static void SaveSettings(T instance, string settingsPath)
            => SaveSettings(instance, settingsPath, "");

        public static string GetSettingFilePath(string settingsPath, string fileId)
            => Path.Combine(settingsPath, typeof(T).Name + fileId + ".xml");
    }
}
